src/PowerShellEditorServices/Language/FindDeclarationVisitor.cs

```diff
@@ -76,19 +78,62 @@ public override AstVisitAction VisitFunctionDefinition(FunctionDefinitionAst fun
-            if (variableExprAst == null ||
-                variableName == null ||
-                !variableExprAst.VariablePath.UserPath.Equals(
-                    variableName,
-                    StringComparison.OrdinalIgnoreCase))
+            if (variableName == null)
             {
                 return AstVisitAction.Continue;
             }
 
-            // TODO also find instances of set-variable
-            FoundDeclaration = new SymbolReference(SymbolType.Variable, variableExprAst.Extent);
-            return AstVisitAction.StopVisit;
+            // The AssignmentStatementAst could contain either of the following Ast types:
+            // VariableExpressionAst, ArrayLiteralAst, ConvertExpressionAst, AttributedExpressionAst
+            // We might need to recurse down the tree to find the VariableExpressionAst we're looking for
+            //                    if (variableExpressionAst.VariablePath.UserPath.Equals(variableName, StringComparison.OrdinalIgnoreCase))
+            FindDeclarationVariableExpressionVisitor visitor = new FindDeclarationVariableExpressionVisitor(symbolRef);
+            assignmentStatementAst.Visit(visitor);
+
+            if (visitor.FoundDeclaration != null)
+            {
+                FoundDeclaration = visitor.FoundDeclaration;
+                return AstVisitAction.StopVisit;
+            }
+            return AstVisitAction.Continue;
+        }
+
+        /// <summary>
+        /// The private visitor used to find the variable expression that matches a symbol
+        /// </summary>
+        private class FindDeclarationVariableExpressionVisitor : AstVisitor
+        {
+            private SymbolReference symbolRef;
+            private string variableName;
+
+            public SymbolReference FoundDeclaration{ get; private set; }
+
+            public FindDeclarationVariableExpressionVisitor(SymbolReference symbolRef)
+            {
+                this.symbolRef = symbolRef;
+                if (this.symbolRef.SymbolType == SymbolType.Variable)
+                {
+                    // converts `$varName` to `varName` or of the form ${varName} to varName
+                    variableName = symbolRef.SymbolName.TrimStart('$').Trim('{', '}');
+                }
+            }
+
+        /// <summary>
+        /// Check if the VariableExpressionAst has the same name as that of symbolRef.
+        /// </summary>
+        /// <param name="variableExpressionAst">A VariableExpressionAst</param>
+        /// <returns>A decision to stop searching if the right VariableExpressionAst was found,
+        /// or a decision to continue if it wasn't found</returns>
+            public override AstVisitAction VisitVariableExpression(VariableExpressionAst variableExpressionAst)
+            {
+                if(variableExpressionAst.VariablePath.UserPath.Equals(variableName, StringComparison.OrdinalIgnoreCase))
+                {
+                    // TODO also find instances of set-variable
```

Comment:
Code for this already exists too: https://github.com/PowerShell/PowerShell/blob/a3786158ca51cd65388743f900b69ec9e253c3d9/src/System.Management.Automation/engine/parser/VariableAnalysis.cs#L38-L311

Reply:
nice link - we'd pretty much need that entire class and rework it to do what we do. I've grabbed a few quick wins from it and applied them.

Reply:
Set-Variable is particularly annoying because it's not an assignmentstatementast

Reply:
Yeah exactly. I think we'll solve 99% with this, no reason to overcomplicate it